added suffix and prefix capabilities in tgocassisstitch

antonhibl opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail including motivation and any context -->

addresses #5125, this allows the user to optionally use either the prefix or
    suffix option when stitching frames together, just use OUTPUTSUFFIX instead
    of OUTPUTPREFIX to use it.

## Related Issue
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

#5125 

## How Has This Been Validated?
<!--- All changes need to be validated to confirm that they produce -->
<!--- scientifically accurate results. -->
<!--- If your changes include any new algorithms or changes to existing algorithms, -->
<!--- please indicate any publications or references for them. -->
<!--- If you manually validated the changes please indicate -->
<!--- what data you used and how you checked the results. -->

Tested locally and works by passing in either a OUTPUTSUFFIX or OUTPUT PREFIX.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation change (update to the documentation; no code change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Infrastructure change (changes to things like CI or the build system that do not impact users)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- - [x] My code follows the code style of this project. -->
- [x] I have read and agree to abide by the [Code of Conduct](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/Code-Of-Conduct.md)
- [x] I have read the [**CONTRIBUTING**](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CONTRIBUTING.md) document.
- [x] My change requires a change to the documentation and I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] I have added myself to the [.zenodo.json](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/.zenodo.json) document.
- [x] I have added any user impacting changes to the [CHANGELOG.md](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CHANGELOG.md) document.

## Licensing
This project is mostly composed of free and unencumbered software released into the public domain, and we are unlikely to accept contributions that are not also released into the public domain. Somewhere near the top of each file should have these words:

> This work is free and unencumbered software released into the public domain. In jurisdictions that recognize copyright laws, the author or authors of this software dedicate any and all copyright interest in the software to the public domain.

- [x] I dedicate any and all copyright interest in this software to the public domain. I make this dedication for the benefit of the public at large and to the detriment of my heirs and successors. I intend this dedication to be an overt act of relinquishment in perpetuity of all present and future rights to this software under copyright law.

Kelvinrr committed:
added findfeaturesSegment script

antonhibl committed:
adding filters for old files in ISISDATA area

antonhibl committed:
adding script changes and fixing PR

antonhibl committed:
used exlcude instead of exclude-from

antonhibl committed:
cleaned up files and script

antonhibl committed:
added test to see if filtering on rclone ISISDATA worked

antonhibl committed:
modified tests

antonhibl committed:
updating imports in tests

antonhibl committed:
added complete test for filters, list of filter files can be changed

antonhibl committed:
updated test to check args rather than files

antonhibl committed:
removed absolute paths

antonhibl committed:
updated test to use exclude list passed in

antonhibl committed:
Updated filter args

I added the following changes to create_rclone_arguments:

- I added a default filter flag --filter with the value of exclude_string

- I added a check for any additional --include and --exclude flags that may have
  been passed in by the user.

- I added logic to merge any additional --include and --exclude flags with the
  default --filter flag. The include and exclude patterns are concatenated with
  the default exclude_string separated by comma.

- I removed the f"--exclude={exclude_string}" from extra_args list, since the
  default exclude_string is now included in the filter flag

- I added the filter flag to the extra_args list using
  extra_args.extend(filter_args)

All of the above changes ensures that the --include and --exclude flags passed
    in by the user are taken into account while creating extra_args, and also
    the logic will merge these flags with the default filter flag, which is the
    recommended way as per the rclone docs.

antonhibl committed:
accidentally deleted a line

antonhibl committed:
Added more logic for when the user provides their own filters

along with a few other changes:

   + I added a check for any --filter flag provided by the user, if present it
    will use it and ignore the default filter flag. Otherwise, it will use the default
    filter flag. This is done to take into account if the user has provided any
    specific filter flag, and it will honor the user's intention of providing
    the filter flag.

   + I added a check for any additional --include and --exclude flags passed in
    by the user and merge them with the filter flag. This is to take into
    account any specific include/exclude patterns that the user wants to apply,
    and merge them with the default filter flag.

   + I added a "+" at the end of the filter string if the user has specified an
    --include flag. This simulates the behavior of --include where it includes
    any patterns specified and excludes everything else.

   + I also added a check for filter_args_provided and if provided, it will use
    this flag, else it will use the default filter flag and merge any additional
    include/exclude flags to it.

There is also a test in the pytest file to check if the filter logic works as
expected, run using `pytest`.

antonhibl committed:
fixed tests and adjusted with new filters

antonhibl committed:
Added several tests to pytests

**The first test that was added is `test_rclone_with_auth`**
    This test is designed to check the behavior of the `rclone` function when it
    is called with an `auth` parameter. This test checks that the `rclone`
    function properly passes the "auth" parameter to the underlying subprocess
    call.

 **The second test that was added is `test_create_rclone_args_with_no_kwargs`**
    This test is designed to check the behavior of the `create_rclone_args`
    function when it is called with no keyword arguments. This test checks that
    the `create_rclone_args` function properly handles the case when it is
    called with no keyword arguments and returns the correct list of arguments
    to be passed to the underlying subprocess call

 **The third test that was added is `test_file_filtering_with_hidden_files`**
    This test is designed to check the behavior of the `file_filtering` function
    when it is called with hidden files in the specified directory. This test
    checks that the `file_filtering` function properly filters out hidden files
    and only returns the non-hidden files in the specified directory.

I have tested to confirm these all run and pass on my machine.

antonhibl committed:
added more tests to pytest

**added `test_rclone` test**
    This test mocks the `subprocess.Popen` function and checks that the output
    of the rclone function matches the expected output when the function is
    invoked with the arguments `lsf`, `test`, `["-l", "-R", "--format", "p",
    "--files-only"]`, `True`, and `True`.

**added `test_rclone_unknown_exception` test**
    This test mocks the `subprocess.Popen` function and checks that the `rclone`
    function raises an exception when an unknown exception is encountered. This
    test uses a mocked class that raises an exception when it is initialized.

I have tested and confirmed these to work on my system.

antonhibl committed:
fixed filter args

    fixed how filters are input, patterns still aren't working I believe I need
    to look at their patterns and ensure the adhere to rclone documentation
    syntactical instructions.

antonhibl committed:
updated filter list,it starts search from {mission_name}/kernels/

antonhibl committed:
finally fixed filters and tests

Kelvinrr committed:
added stuff

Kelvinrr committed:
fixed parsing issues

Kelvinrr committed:
cleaning up

Kelvinrr committed:
more cleaning up

antonhibl committed:
added suffix and prefix capabilities in tgocassisstitch

    addresses #5125, this allows the user to optionally use either the prefix or
    suffix option when stitching frames together, just use OUTPUTSUFFIX instead
    of OUTPUTPREFIX to use it.

antonhibl committed:
added suffix and prefix capabilities in tgocassisstitch

    addresses #5125, this allows the user to optionally use either the prefix or
    suffix option when stitching frames together, just use OUTPUTSUFFIX instead
    of OUTPUTPREFIX to use it.

antonhibl committed:
minor fixes

antonhibl committed:
minor fixes

antonhibl committed:
typo

antonhibl committed:
typo

antonhibl committed:
more typos

antonhibl committed:
more typos

antonhibl committed:
close to done, one error left

    need to figure out a replacement for ui.IsOptionSet as that is not actually
    a function. I need another way to test CLLI args.

antonhibl committed:
close to done, one error left

    need to figure out a replacement for ui.IsOptionSet as that is not actually
    a function. I need another way to test CLLI args.

antonhibl committed:
final fixes to implement ui changes

antonhibl committed:
final fixes to implement ui changes

antonhibl committed:
fixed

    now employs a boolean flag to decide if the name parsed with out is a suffix
    or prefix, it defaults to a prefix.

antonhibl committed:
fixed

    now employs a boolean flag to decide if the name parsed with out is a suffix
    or prefix, it defaults to a prefix.

antonhibl committed:
adding some tests to confirm output runs as expected.

antonhibl committed:
adding some tests to confirm output runs as expected.

antonhibl committed:
adding to changelog

antonhibl committed:
adding to changelog

antonhibl committed:
Merge branch 'dev' into tgo-cassis-stitching

AustinSanders reviewed (commented):
I'm concerned that these changes might break existing pipelines that rely on consistent behavior with respect to parameter names and behaviors.

Ideally, this PR would preserve the integrity of the current naming mechanism and add changes that are optional.

antonhibl committed:
Merge branch 'dev' into tgo-cassis-stitching

antonhibl committed:
modified to allow prefix, suffix, both, or neither for better extensibility

    it also maintains original value names so older scripts and pipelines will
    retain functionality.

antonhibl committed:
modified to allow prefix, suffix, both, or neither for better extensibility

    it also maintains original value names so older scripts and pipelines will
    retain functionality.

antonhibl committed:
adds another optional arg, cubename

    lets the user specify an optional cubename parameter which replaces the
    timestamp style naming for the output stitched cube. If the parameter is not
    given it defaults to the timestamp style naming convention for retaining
    existing functionality.

antonhibl committed:
adds another optional arg, cubename

    lets the user specify an optional cubename parameter which replaces the
    timestamp style naming for the output stitched cube. If the parameter is not
    given it defaults to the timestamp style naming convention for retaining
    existing functionality.

antonhibl committed:
added details to changelog

antonhibl committed:
added details to changelog

antonhibl committed:
removed debugging code

antonhibl committed:
removed debugging code

antonhibl committed:
added changelog notes to xml for tgocassisstitch

antonhibl committed:
added changelog notes to xml for tgocassisstitch

antonhibl committed:
typo in changelog

antonhibl committed:
typo in changelog

AustinSanders commented on isis/src/tgo/apps/tgocassisstitch/tgocassisstitch.cpp:
```diff
@@ -66,20 +66,69 @@ namespace Isis {
-        FileName frameFileName(outputBaseName + "-" + frameIdentifier + ".cub");
-        stitchFrame( frameMap.values(frameKey), frameFileName );
-        stitchProgress.CheckStatus();
+        if ((outputPrefix != "nil") && (outputSuffix == "nil")) {
```
We should attempt to minimize the number of logical branches here.  Instead of having a branch for each combination of prefix / suffix, can we use something like:

```
if(outputPrefix == "nil"){
  outputPrefix = "";
}
if (outputSuffix == "nil"){
  outputSuffix = "";
}
```
Then we would just need conditionals for frameidentifier vs cubename, and we could get rid of all the nested logic e.g.:
```
FileName frameFileName(outputPrefBaseName + frameletCubeName + outputSuffBaseName ".cub");
```

I know this will get rid of the '-', but that's something that the user could easily control by specifying the separator as part of the prefix/suffix.  This might even be viewed as a "feature" since the user could then specify a '.' separator if they desire.  It's somewhat common to see something like mydata.proj.cub.

antonhibl committed:
fixes from review

antonhibl committed:
fixes from review

acpaquette commented:
There is also a failing test case that needs to be update from:
https://jenkins-hsts.prod-asc.chs.usgs.gov/job/ISIS/view/change-requests/job/PR-5132/

The specific test is `tgocassisstitch_app_test_default`

antonhibl committed:
added cubename argument logic

antonhibl committed:
added cubename argument logic

antonhibl committed:
typo in else statement

antonhibl committed:
typo in else statement

antonhibl committed:
modifying tests

antonhibl committed:
modifying tests

antonhibl committed:
removed faulty/old tests based on makefiles

antonhibl committed:
removed faulty/old tests based on makefiles

antonhibl reviewed (commented):
should be addressed with the new logic I added

AustinSanders commented:
I'm still seeing test failures for:
	2073 - TgoCassisstitch.TgoCassisstitchMultiframeTest (Failed)
	2074 - TgoCassisstitch.TgoCassisstitchSingleframeTest (Failed)
	2309 - TgoCassisModuleKernels.TgoCassisStitchUnstitch (Failed)
	2314 - TgoCassisModuleKernels.TgoCassisTestProjSingleStitchedFrame (Failed)

Most of the errors are something like ``` **I/O ERROR** Unable to open [/tmp/qt_temp-ftZxne/CAS-MCO-2016-11-22T16:16:10.833.cub].``` I'm wondering if there's just a difference in the output filename?  You can run the tgocassisstitch tests locally with ``` ctest -VV -R TgoCassissitch* ```, and that should give you a lot more info about the test failures.

EDIT:

It looks like Jenkins got disconnected as a result of the transition to DGEC, and it was hitting the wrong URL.  I'm wondering if this didn't pick up the code changes and just used a cached version?  I'll rerun CI for this and check back

antonhibl commented:
@AustinSanders sounds good, just let me know

AustinSanders commented:
@antonhibl Have you run the tests locally?  They're still failing on Jenkins, but it's also showing that no changes were detected since the last build.  The logs show that it's pulling from the correct repo.

antonhibl committed:
fixed single and multi frame tests for tgocassisstitch

antonhibl committed:
fixed single and multi frame tests for tgocassisstitch

antonhibl commented:
@AustinSanders these tests are fixed and this should be good to merge.

antonhibl reviewed (commented):
should be addressed

antonhibl committed:
Merge remote-tracking branch 'upstream/dev' into tgo-cassis-stitching

jrcain-usgs committed:
Updated Qmos to load planetographic projections without loading planetocentric first (#5130)

* Updated Grid-specific lat-lons to track proj radii

* updated changelog with qmos fix

acpaquette commented:
@antonhibl This is still failing on jenkins, see https://jenkins-hsts.prod-asc.chs.usgs.gov/job/ISIS/view/change-requests/job/PR-5132/

Per the build on March 15th, running another just to confirm

AustinSanders commented:
I pulled this and ran tests locally, and I'm getting 
```
The following tests FAILED:
	2309 - TgoCassisModuleKernels.TgoCassisStitchUnstitch (Failed)
	2314 - TgoCassisModuleKernels.TgoCassisTestProjSingleStitchedFrame (Failed)
```

antonhibl committed:
adding to changelog

antonhibl committed:
modified to allow prefix, suffix, both, or neither for better extensibility

    it also maintains original value names so older scripts and pipelines will
    retain functionality.

antonhibl committed:
modified to allow prefix, suffix, both, or neither for better extensibility

    it also maintains original value names so older scripts and pipelines will
    retain functionality.

antonhibl committed:
added details to changelog

antonhibl committed:
added details to changelog

antonhibl committed:
changed string logic, fixed issues with tests I think.

antonhibl committed:
changed string logic, fixed issues with tests I think.

Kelvinrr commented on CHANGELOG.md:
```diff
@@ -36,14 +36,50 @@ release.
 ## [Unreleased]
 
 ### Changed
+<<<<<<< HEAD
+<<<<<<< HEAD
+- Modified tgocassisstitch to optionally allow either a outputprefix or an
+  outputsuffix, both, or neither for naming convention purposes [#5125](https://github.com/USGS-Astrogeology/ISIS3/issues/5125)
+=======
+<<<<<<< HEAD
```
Looks like merge conflict markers need to be removed.

Kelvinrr commented on isis/src/tgo/apps/tgocassisstitch/tgocassisstitch.cpp:
```diff
@@ -66,20 +66,71 @@ namespace Isis {
-        FileName frameFileName(outputBaseName + "-" + frameIdentifier + ".cub");
-        stitchFrame( frameMap.values(frameKey), frameFileName );
-        stitchProgress.CheckStatus();
+<<<<<<< HEAD
+<<<<<<< HEAD
+        if(outputPrefix == "nil"){
+          outputPrefix = "";
+        }
+        if (outputSuffix == "nil"){
+          outputSuffix = "";
+        }
+=======
+>>>>>>> 37c1659f8 (changed string logic, fixed issues with tests I think.)
```
Not gonne mark all of them, but they are in pretty much all files.

antonhibl committed:
fixed headers from git rebase that accidentally got saved in files

antonhibl committed:
Merge branch 'dev' into tgo-cassis-stitching

antonhibl committed:
more git header removals

antonhibl committed:
removed random git headers again

    no idea why tf those were added but that was irritating.

antonhibl committed:
Merge branch 'dev' into tgo-cassis-stitching

antonhibl commented:
@Kelvinrr knocking on wood, these should be fixed. I had some kind of thing happen where those got inserted by gitkraken. I still need to debug the TgoCassisModuleTests though.

acpaquette commented:
@antonhibl Good on building now, only one more to go! https://jenkins-hsts.prod-asc.chs.usgs.gov/job/ISIS/view/change-requests/job/PR-5132/

The DemCube test failure is something else you don't have to worry about

antonhibl committed:
fixing merge conflict tags from cherry picking

antonhibl committed:
Merge branch 'tgo-cassis-stitching' into TGOv2

antonhibl committed:
Merge pull request #1 from antonhibl/TGOv2

Tgo fixes, that PR was a mess lol

antonhibl committed:
resolved more tests, 1 to go

    the test left is a make test so need to look into it more.

antonhibl committed:
Merge pull request #2 from antonhibl/TGOv2

resolved more tests, 1 to go

antonhibl committed:
Merge branch 'DOI-USGS:dev' into tgo-cassis-stitching

antonhibl commented:
closed this in favor of opening #5162 with a cleaner commit history and branch.
